dipy/reconst/ivim.py

```diff
@@ -0,0 +1,478 @@
+
+if SCIPY_LESS_0_17:
+    from scipy.optimize import leastsq
+else:
+    from scipy.optimize import least_squares
+
+
+def ivim_prediction(params, gtab, S0=1.):
+    """The Intravoxel incoherent motion (IVIM) model function.
+
+    Parameters
+    ----------
+    params : array
+        An array of IVIM parameters - [S0, f, D_star, D]
+
+    gtab : GradientTable class instance
+        Gradient directions and bvalues
+
+    S0 : float, optional
+        This has been added just for consistency with the existing
+        API. Unlike other models, IVIM predicts S0 and this is over written
+        by the S0 value in params.
+
+    """
+    S0, f, D_star, D = params
+    b = gtab.bvals
+    S = S0 * (f * np.exp(-b * D_star) + (1 - f) * np.exp(-b * D))
+    return S
+
+
+def _ivim_error(params, gtab, signal, scaling=np.array([1., 1., 1., 1.])):
+    """Error function to be used in fitting the IVIM model.
+
+    Parameters
+    ----------
+    params : array
+        An array of IVIM parameters. [S0, f, D_star, D]
+
+    gtab : GradientTable class instance
+        Gradient directions and bvalues.
+
+    signal : array
+        Array containing the actual signal values.
+
+    scaling : array
+        An array to scale the various parameters.
+        default : [1., 1., 1., 1.]
+
+    """
+    return (signal * scaling[0]) - ivim_prediction(params * scaling, gtab)
+
+
+def f_D_star_prediction(params, gtab, x0):
+    """Function used to predict f and D_star when S0 and D are known.
+
+    This restricts the value of 'f' and 'D_star' to physically feasible
+    solutions since a direct fitting leads to 'f' jumping to very high
+    values and 'D_star' vanishing.
+
+    Parameters
+    ----------
+    params : array, dtype=float
+        An array containg the values of f and D_star.
+
+    gtab : GradientTable class instance
+        Gradient directions and bvalues.
+
+    other_params : array, dtype=float
+        The parameters S0 and D which are obtained from a linear fit.
+    """
+    f, D_star = params
+    S0, D = x0[0], x0[3]
+    b = gtab.bvals
+    S = S0 * (f * np.exp(-b * D_star) + (1 - f) * np.exp(-b * D))
+    return S
+
+
+def f_D_star_error(params, gtab, signal, x0):
+    """Error function used to fit f and D_star keeping S0 and D fixed
+
+    Parameters
+    ----------
+    params : array
+        The value of f and D_star.
+
+    gtab : GradientTable class instance
+        Gradient directions and bvalues.
+
+    signal : array
+        Array containing the actual signal values.
+
+    other_params : array
+        The parameters S0 and D which are fixed and obtained from a linear fit.
+    """
+    f, D_star = params
+    S0, D = x0[0], x0[3]
+
+    return signal - f_D_star_prediction([f, D_star], gtab, x0)
+
+
+class IvimModel(ReconstModel):
+    """Ivim model
+    """
+
+    def __init__(self, gtab, split_b_D=400.0, split_b_S0=200., bounds=None,
+                 tol=1e-15, scaling=np.array([10e-4, 10e2, 1000., 10000.]),
```

Comment:
I think f should not be scaled at all and the factor should be 1. But if I do that I get test failures for all parameters. I think its better to handle scaling with least_squares built in method instead of this way.